libsolidity/analysis/PostTypeContractLevelChecker.cpp

```diff
@@ -144,4 +144,47 @@ void PostTypeContractLevelChecker::checkStorageLayoutSpecifier(ContractDefinitio
 			baseSlotExpression.location(),
 			"Contract extends past the end of storage when this base slot value is specified."
 		);
+
+	warnStorageLayoutBaseNearStorageEnd(_contract);
+}
+
+void PostTypeContractLevelChecker::warnStorageLayoutBaseNearStorageEnd(ContractDefinition const& _contract)
+{
+	if (Error::containsErrors(m_errorReporter.errors()))
+		return;
+
+	auto storageSize = u256(contractStorageSizeUpperBound(_contract, VariableDeclaration::Location::Unspecified));
+	auto baseSlot = layoutBaseForInheritanceHierarchy(_contract, DataLocation::Storage);
+	if (
+		u256 slotsLeft = std::numeric_limits<u256>::max() - baseSlot - storageSize;
```

Comment:
This should have an assert, because `u256` will truncate if the value is out of range.

Also, it's pretty important to know where you have `u256` and where `bigint`, so `auto` here obscures really useful information.